src/lib.rs

```diff
@@ -54,77 +46,54 @@ pub fn get_zones_and_stats(
-            let level = relation
-                .tags
-                .get("admin_level")
-                .and_then(|s| s.parse().ok());
-
-            // Skip administrative region without name
-            let name = match relation.tags.get("name") {
-                Some(val) => val,
-                None => {
-                    warn!(
-                        "relation/{}: adminstrative region without name, skipped",
-                        relation.id.0
-                    );
-                    continue;
+        if let OsmObj::Relation(ref relation) = *obj {
+            if let Some(zone) = zone::Zone::from_osm_with_geom(relation, &objects) {
+                // Ignore zone without boundary polygon
+                if zone.boundary.is_some() {
+                    stats.process(&zone);
+                    zones.push(zone);
                 }
-            };
+            }
+        }
+    }
 
-            let coord_center = relation
-                .refs
-                .iter()
-                .find(|r| r.role == "admin_centre")
-                .and_then(|r| objects.get(&r.member))
-                .and_then(|o| o.node())
-                .map(|node| mimir::Coord::new(node.lat(), node.lon()));
-            let zip_code = relation
-                .tags
-                .get("addr:postcode")
-                .or_else(|| relation.tags.get("postal_code"))
-                .map_or("", |val| &val[..]);
-            let zip_codes = zip_code
-                .split(';')
-                .filter(|s| !s.is_empty())
-                .map(|s| s.to_string())
-                .sorted();
-            let boundary = build_boundary(relation, &objects);
-            let zone = zone::Zone {
-                id: relation.id.0.to_string(),
-                admin_level: level,
-                admin_type: None,
-                name: name.to_string(),
-                zip_codes: zip_codes,
-                center: coord_center.unwrap_or_else(|| make_centroid(&boundary)),
-                boundary: boundary,
-                parent: None,
-                tags: vec![],
-            };
+    return Ok((zones, stats));
+}
 
-            // Ignore zone without boundary polygon
-            if zone.boundary.is_none() {
-                continue;
-            }
+pub fn get_zones_and_stats_without_geom(
+    pbf: &mut OsmPbfReader<File>,
+) -> Result<(Vec<zone::Zone>, CosmogonyStats), Error> {
+    info!("Reading pbf without geometries...");
 
-            zone.admin_level.map(|level| {
-                let count = stats.level_counts.entry(level).or_insert(0);
-                *count += 1;
-            });
-            zones.push(zone);
+    let mut zones = vec![];
+    let mut stats = CosmogonyStats::default();
+
+    for obj in pbf.par_iter().map(Result::unwrap) {
+        if !is_admin(&obj) {
+            continue;
+        }
+        if let OsmObj::Relation(ref relation) = obj {
+            if let Some(zone) = zone::Zone::from_osm(relation) {
```

Comment:
wouldn't it be easier to just pass the `parse_geom` boolean here ? (so only one `get_zones` method, less code dupplicate

Reply:
The problem is that I want to avoid calling `get_objs_and_deps`, I could pass an option for `objects`. Do you think it would work?

Reply:
ah, my bad, I didn't saw there was a difference here too.